wire-ios-data-model/Source/MLS/CoreCryptoProvider.swift

```diff
@@ -49,30 +58,90 @@ public class CoreCryptoProvider: CoreCryptoProviderProtocol {
         self.cryptoboxMigrationManager = cryptoboxMigrationManager
     }
 
-    // NOTE: this will turn async when we upgrade CC
-    public func coreCrypto(requireMLS: Bool = false) throws -> SafeCoreCryptoProtocol {
-        // TODO: this lock should go away when this function turn async and the class can become an actor
-        return try lock.withLock {
-            let coreCrypto = if let coreCrypto = coreCrypto {
-                coreCrypto
-            } else {
-                try createCoreCrypto()
+    public func coreCrypto(requireMLS: Bool = false) async throws -> SafeCoreCryptoProtocol {
+        let coreCrypto = try await getCoreCrypto()
+
+        if requireMLS {
+            try await initialiseMLS(coreCrypto: coreCrypto)
+        }
+
+        return coreCrypto
+    }
+
+    // Initialise MLS with guranteees that only one task is performing
+    // the operation while others wait for it to complete.
+    //
+    // Based on the structured caching in an actor:
+    // https://forums.swift.org/t/structured-caching-in-an-actor/65501/13
+    private func initialiseMLS(coreCrypto: SafeCoreCrypto) async throws {
+        guard !initialisatingMls else {
+            return try await withCheckedThrowingContinuation { continuation in
+                initialiseMlsContinuations.append(continuation)
             }
+        }
 
-            self.coreCrypto = coreCrypto
+        do {
+            let provider = CoreCryptoConfigProvider()
```

Comment:
```suggestion
            initialisatingMLS = true
            let provider = CoreCryptoConfigProvider()
```
`initialisatingMLS` is never set to `true`.
Also I'd prefer spelling MLS in capitals.

Reply:
Done